### PR TITLE
rustc: 1.21.0 -> 1.22.1

### PR DESCRIPTION
**Pull request by zimbatm**
###### Motivation for this change

duh

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by zimbatm** [comment]
can anyone test it on macOS?

**Comment by grahamc** [comment]
@GrahamcOfBorg rustc

**Comment by zimbatm** [comment]
@grahamcofborg eval

**Comment by grahamc** [comment]
No need to call eval manually, unless something went wrong mysteriously with the first one :)

**Comment by zimbatm** [comment]
alright :)

**Comment by zimbatm** [comment]
So cargo-vendor is supposed to support git dependencies now. I have https://github.com/zimbatm/nixpkgs/tree/alacritty-git-deps to demonstrate the concept but there is some complication. For some reason it still needs a registry:
> error: no source URL specified for `source.https://github.com/jwilm/libfontconfig`, need either `registry` or `local-registry` defined

**Comment by LnL7** [comment]
@zimbatm Let me know if you want help for the darwin changes.

**Comment by zimbatm** [comment]
@LnL7 if you could nox-review on Darwin it would be great!

**Comment by Ma27** [comment]
@zimbatm not sure if I got your comment right, but is there any way to fix the resolution issue with git dependencies in Cargo? I tried to package uutils/coreutils for @NeQuissimus once and I failed because of such Cargo problems...

**Comment by Mic92** [comment]
@Ma27 I am not sure what `cargo-vendor` does at the moment, but according to the documentation, should write an source replacement entry like this: https://github.com/alexcrichton/cargo/blob/5b08b8fe1019147fe489db17a9a8ae7ebe97f9e9/src/doc/source-replacement.md#configuration

**Comment by orivej** [comment]
@zimbatim Why do you mark `bokeh` as broken? It builds on Linux: https://hydra.nixos.org/jobset/nixpkgs/trunk/jobs-tab?filter=bokeh

**Comment by orivej** [comment]
rustc fails to build on Darwin:
```
configuring                                    
configure flags: --prefix=/nix/store/sg80860pldx9g6d8awfdxidl17l9yq92-rustc-1.22.1 --release-channel=stable --enable-local-rust --local-rust-root=/nix/store/d5fxqy76rqscy
aw9p1mnngkz86wq8pv2-rustc-bootstrap-1.21.0 --enable-rpath --enable-vendor --disable-locked-deps --default-linker=/nix/store/4z8vhlqrnfpjr963pgxhccyzhgpjflw7-clang-wrapper
-4.0.1/bin/cc --default-ar=/nix/store/09jllpa592yk1yiz018yv6s5jjmzymf0-cctools-binutils-darwin/bin/ar --enable-clang
configure: processing command line                                
configure:                                        
configure: error: Option '--enable-clang' is not recognized     
builder for ‘/nix/store/6m5wrj6bhzy2kz6lb70fbq3wws8chj74-rustc-1.22.1.drv’ failed with exit code 1
```

**Comment by zimbatm** [comment]
@orivej for bokeh, I am getting the following error at the check phase:
```
  File "bokeh/models/annotations.py", line 15, in <module>
    from ..core.property_mixins import FillProps, LineProps, TextProps
ImportError: cannot import name FillProps
```

**Comment by orivej** [comment]
@zimbatm Are you building on NixOS or other Linux? With or without sandboxing? Could you attach the complete build log? Nevertheless, since bokeh builds on Hydra (and my machines), it should not be disabled in this PR. (I'm also trying to restore the support for Darwin by fixing the build of python.pkgs.selenium.)

To fix the build phase on Darwin, delete the line with `"--enable-clang"`. Then the build phase succeeds, but one of the tests fails… (I've confirmed that rust 1.21.0 builds fine on the same machine.)

File: pkgs/development/python-modules/bokeh/default.nix [hunk 1]
```diff
@@ -75,6 +75,7 @@ buildPythonPackage rec {
   '';
 
   meta = {
+    broken = true;
```
Comment (FRidh) [inline pkgs/development/python-modules/bokeh/default.nix]: please remove this. As @orivej pointed out, it builds, and if it doesn't, it needs to be fixed outside this PR.

**Comment by zimbatm** [comment]
@Ma27 let's discuss the cargo git situation in the next PR, once rust itself has been updated. Basically it's not working either after the update so it will need to be investigated.

@orivej removed `--enable-clang`, can you give me the next failure?

@orivej @FRidh re-enabled bokeh but it's still failing for me on NixOS with sandbox enabled.

**Comment by Ma27** [comment]
@zimbatm great! Just feel free to ping me if there's anything I could do to fix this cargo issue :)

**Comment by orivej** [comment]
Here are my two attempts to build rust 1.22.1 on Darwin: [rustc-1.22.1-try1.log](https://github.com/NixOS/nixpkgs/files/1503455/rustc-1.22.1-try1.log), [rustc-1.22.1-try2.log](https://github.com/NixOS/nixpkgs/files/1503456/rustc-1.22.1-try2.log). For comparison, the log of building rust 1.21.0: [rustc-1.21.0.log](https://github.com/NixOS/nixpkgs/files/1503454/rustc-1.21.0.log).

**Comment by orivej** [comment]
> re-enabled bokeh but it's still failing for me on NixOS with sandbox enabled (EDIT: actually it's also failing on master for me). There is no maintainer for the package so either become a maintainer or I will mark it as broken.

You should not mark the package as broken when you do not know why it is breaking and it works for everyone but you. I have already posted the [link](https://hydra.nixos.org/jobset/nixpkgs/trunk/jobs-tab?filter=bokeh) that proves that Hydra can build it on master. Now I have [added](https://github.com/NixOS/nixpkgs/commit/c97ea968609d30e897dff4fd0d199468132526c0) myself to its maintainers. Please open a separate issue and attach the complete build log.

**Comment by zimbatm** [comment]
@GrahamcOfBorg build rustc

**Comment by LnL7** [comment]
That's a sandboxing issue with the bootstrap rust.

**Comment by zimbatm** [comment]
@LnL7 are you going to look into it? The other platforms it just a timeout because rust takes a long time to compile.

The last change I pushed allows for git dependencies in the Cargo.lock, but it will mean updating all the cargoSha256 again. I will do it but I want to get rust sorted out first.

**Comment by LnL7** [comment]
I'm looking at the sandbox impurity, but I'm fine with merging this if it works without.

**Comment by orivej** [comment]
Rust 1.21.0 has one test that occasionally fails on darwin (`run-make/issue-26092`: https://hydra.nixos.org/job/nixpkgs/trunk/rustc.x86_64-darwin/all). Rust 1.22.1 has three failing tests and IIRC at least two of them always fail on darwin (`run-make/issue-26092` is not one of the three).

**Comment by LnL7** [comment]
Sorry for the long delay. I fixed the sandboxing issues on master and pushed a fixed the 1.22.1 build.

**Comment by globin** [comment]
@GrahamcOfBorg build rustc

**Comment by zimbatm** [comment]
@globin all the cargoSha256 need to be fixed now. You can revert 71064ac and I will make another PR for the git support.

**Pull request merged**